x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/sp/SamlServiceProviderResolver.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.idp.saml.sp;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.cache.Cache;
+import org.elasticsearch.common.cache.CacheBuilder;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.iterable.Iterables;
+import org.elasticsearch.xpack.idp.privileges.ServiceProviderPrivileges;
+import org.elasticsearch.xpack.idp.saml.idp.SamlIdentityProvider;
+import org.elasticsearch.xpack.idp.saml.sp.SamlServiceProviderIndex.DocumentSupplier;
+import org.elasticsearch.xpack.idp.saml.sp.SamlServiceProviderIndex.DocumentVersion;
+import org.joda.time.ReadableDuration;
+import org.opensaml.security.x509.BasicX509Credential;
+import org.opensaml.security.x509.X509Credential;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class SamlServiceProviderResolver {
+
+    private static final int CACHE_SIZE_DEFAULT = 1000;
+    private static final TimeValue CACHE_TTL_DEFAULT = TimeValue.timeValueMinutes(60);
+
+    public static final Setting<Integer> CACHE_SIZE
+        = Setting.intSetting("xpack.idp.sp.cache.size", CACHE_SIZE_DEFAULT, Setting.Property.NodeScope);
+    public static final Setting<TimeValue> CACHE_TTL
+        = Setting.timeSetting("xpack.idp.sp.cache.ttl", CACHE_TTL_DEFAULT, Setting.Property.NodeScope);
+
+    private final Cache<String, CachedServiceProvider> cache;
+    private final SamlServiceProviderIndex index;
+    private final SamlIdentityProvider identityProvider;
+
+    public SamlServiceProviderResolver(Settings settings, SamlServiceProviderIndex index, SamlIdentityProvider identityProvider) {
+        this.cache = CacheBuilder.<String, CachedServiceProvider>builder()
+            .setMaximumWeight(CACHE_SIZE.get(settings))
+            .setExpireAfterAccess(CACHE_TTL.get(settings))
+            .build();
+        this.index = index;
+        this.identityProvider = identityProvider;
+    }
+
+    /**
+     * Find a {@link SamlServiceProvider} by entity-id.
+     *
+     * @param listener Callback for the service provider object. Calls {@link ActionListener#onResponse} with a {@code null} value if the
+     *                 service provider does not exist.
+     */
+    public void resolve(String entityId, ActionListener<SamlServiceProvider> listener) {
```

Comment:
Mostly because the alternative needs more plumbing.
It needs:
1. A transport action to clear the cache on change
2. something to fix the cache if a node leaves the cluster for a period of time, in case it missed clear-cache events
3. A rest action to clear the cache if you've done something that wasn't automatically detected as a change
4. docs explaining the conditions under which you'd need to call the rest action

Our history with doing that for users & roles has shown that issues slip through and they're really hard to track down.

Reply:
Thanks for the explanation - I agree with your reasoning and it's good to have it written down for future reference too